src/pad/src/ICeWall.cpp

```diff
@@ -0,0 +1,1070 @@
+                               "IO_CORNER_NORTH_EAST",
+                               "IO_CORNER_SOUTH_WEST",
+                               "IO_CORNER_SOUTH_EAST"}) {
+    odb::dbRow* row;
+    if (ring_index >= 0) {
+      row = findRow(fmt::format("{}_{}", row_name, ring_index));
+    } else {
+      row = findRow(row_name);
+    }
+    if (row == nullptr) {
+      continue;
+    }
+
+    const std::string corner_name = fmt::format("{}_INST", row->getName());
+    auto* inst = odb::dbInst::create(block, master, corner_name.c_str());
+
+    const odb::Rect row_bbox = row->getBBox();
+
+    inst->setOrient(row->getOrient());
+    inst->setLocation(row_bbox.xMin(), row_bbox.yMin());
+    inst->setPlacementStatus(odb::dbPlacementStatus::FIRM);
+  }
+}
+
+void ICeWall::placePad(odb::dbMaster* master,
+                       const std::string& name,
+                       odb::dbRow* row,
+                       int location,
+                       bool mirror)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  odb::dbInst* inst = block->findInst(name.c_str());
+  if (inst == nullptr) {
+    if (master == nullptr) {
+      return;
+    }
+    inst = odb::dbInst::create(block, master, name.c_str());
+  }
+
+  if (row == nullptr) {
+    return;
+  }
+
+  odb::dbTransform orient(odb::dbOrientType::R0);
+  if (mirror) {
+    const odb::dbTransform mirror_transform(odb::dbOrientType::MY);
+    orient.concat(mirror_transform);
+  }
+
+  placeInstance(row, snapToRowSite(row, location), inst, orient.getOrient());
+}
+
+int ICeWall::snapToRowSite(odb::dbRow* row, int location) const
+{
+  int x, y;
+  row->getOrigin(x, y);
+  const odb::Point origin(x, y);
+
+  const double spacing = row->getSpacing();
+  int relative_location;
+  if (row->getDirection() == odb::dbRowDir::HORIZONTAL) {
+    relative_location = location - origin.x();
+  } else {
+    relative_location = location - origin.y();
+  }
+
+  int site_count = std::round(relative_location / spacing);
+  site_count = std::max(0, site_count);
+  site_count = std::min(site_count, row->getSiteCount());
+
+  return site_count;
+}
+
+odb::dbRow* ICeWall::findRow(const std::string& name) const
+{
+  for (auto* row : getBlock()->getRows()) {
+    if (row->getName() == name) {
+      return row;
+    }
+  }
+  return nullptr;
+}
+
+void ICeWall::placeInstance(odb::dbRow* row,
+                            int index,
+                            odb::dbInst* inst,
+                            odb::dbOrientType base_orient) const
+{
+  const int origin_offset = index * row->getSpacing();
+
+  const odb::Rect row_bbox = row->getBBox();
+  const std::string row_name = row->getName();
+
+  odb::dbTransform xform(base_orient);
+  xform.concat(row->getOrient());
+  inst->setOrient(xform.getOrient());
+  const odb::Rect inst_bbox = inst->getBBox()->getBox();
+
+  odb::Point index_pt;
+  if (row_name.find("NORTH") != std::string::npos) {
```

Comment:
Can we avoid using string lookup to find the side?

Reply:
If you have suggestions, I'm open. The reason for the string check was it made things easy to read and works.